### PR TITLE
[Windows] Adds a native windows filesystem implementation

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Commit Message:

Refactor `filesystem_impl` to use win32 APIs instead of POSIX subsystem. Fixes issue #11655.

Additional Description:

For easier review the major changes are the following:

1. Adds a Win32 impl
1. Strips away the `file_shared_impl` because it relied on `errno` for error handling. This change cascades to the POSIX implementation
1. Adds a preliminary error mapping so we don't rely on error strings.

Risk Level: Low
Testing: Already tested in `filesystem_impl_test`
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by davinci26** [comment]
cc: @envoyproxy/windows-dev

gimme one CI run to clean up any formatting or cascading build errors

File: source/common/filesystem/win32/filesystem_impl.cc
Comment (nigriMSFT) [inline source/common/filesystem/win32/filesystem_impl.cc:35]: Lets try to use 0 instead of NULL for the DWORD parameters

Comment (davinci26) [reply source/common/filesystem/win32/filesystem_impl.cc:35]: Sure, will update. Out of curiosity what is the reason behind that?

Comment (nigriMSFT) [reply source/common/filesystem/win32/filesystem_impl.cc:35]: People usually expect NULL for pointers and not integers. If this repo has a style rule against that then by all means follow the repo style

Comment (davinci26) [reply source/common/filesystem/win32/filesystem_impl.cc:35]: Gotcha. Thanks for the explanation.

Comment (davinci26) [reply source/common/filesystem/win32/filesystem_impl.cc]: Fixed in e22ac3eaa32da79d7ce14dc1a0045ceb28fd4891

File: test/common/filesystem/filesystem_impl_test.cc
Comment (wrowe) [inline test/common/filesystem/filesystem_impl_test.cc:351]: I'm wondering if we want to ensure these have the same observed error in the windows and unix implementations?

File: include/envoy/common/platform.h
Comment (wrowe) [inline include/envoy/common/platform.h:64]: The os_h_ is a somewhat awkward name imo, at least 'fd' is a well recognized Posix construct. At minimum, os_handle_t, but that's a foreign concept to Posix.

What I'm wondering, because a SOCKET is simply a special case of a HANDLE, is whether we need two types different types for Windows, or whether we can simply continue to use os_fd_t?

In any case, I don't think we want to be jumping between the typedef and using constructions.

**Review by wrowe — COMMENTED** [review]
A couple quick comments for discussion, but generally looking really good here, thanks!

File: test/common/filesystem/filesystem_impl_test.cc
Comment (davinci26) [reply test/common/filesystem/filesystem_impl_test.cc:351]: This is my initial goal but the thing that kept from doing it is that the error codes that the platform codes are so different. 

The error in POSIX is `EBADF` which is a really generic error so it's hard to map it back to the more specific error. Alternatively, we could make the Windows error more generic but then we would be "losing" information.

File: include/envoy/common/platform.h
Comment (davinci26) [reply include/envoy/common/platform.h:64]: 1. For the `using` vs `typdef` I tried this because `clang-tidy` raises an error with typedef. The error is prefer `using` over typedef. I tried using as an alternative but this did not fix the issue.

1. I can change it to `os_handle_t `. From what I understood while I was reading about these two types I could not find a way to have them as a common abstraction. At least without some extra assumptions or some 'casting'. The reason I went for two different types is because the use of `os_handle_t` is limited to the filesystem implementation.

Comment (davinci26) [reply include/envoy/common/platform.h]: @wrowe this is the clang-tidy error. Do you know what a good resolution would be?

Comment (wrowe) [reply include/envoy/common/platform.h]: @htuch or @mattklein123 any thoughts on using vs typedef in this context?

Comment (wrowe) [reply include/envoy/common/platform.h]: AIUI they are #defined together on windows. Since every socket is a handle, every linux socket is an fd, I'd rather have a resolution at os_fd_t vs os_socket_fd_t if you want two distinct types.

Comment (mattklein123) [reply include/envoy/common/platform.h]: There is typically a way via NOLINT to stop tidy complaining. See other examples.

Comment (davinci26) [reply include/envoy/common/platform.h]: Both of the issues are fixed in 47371422958b974e7589af88e931c08f9822c8e1

**Comment by davinci26** [comment]
Attempting to bump coverage above threshold. Technically it shouldn't have been affected by the change since the code is a null-op in linux.

**Comment by davinci26** [comment]
`clang-tidy` fails due to #13162 that is coming from upstream

**Review by wrowe — APPROVED** [review]
LGTM!

**Comment by davinci26** [comment]
@envoyproxy/senior-maintainers can I also get a review from you.

**Review by mattklein123 — APPROVED** [review]
Nice cleanup. Thanks!

**Pull request merged**
